app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt

```diff
@@ -863,6 +880,108 @@ class VaultRepositoryImpl(
             )
     }
 
+    @Suppress("LongMethod")
+    override suspend fun registerFido2Credential(
+        fido2CredentialRequest: Fido2CredentialRequest,
+        selectedCipherView: CipherView,
+        isVerificationSupported: Boolean,
+        checkUser: suspend (CheckUserOptions, UiHint?) -> CheckUserResult,
+    ): Fido2CreateCredentialResult {
+        val userId = activeUserId ?: return Fido2CreateCredentialResult.Error(
+            CreateCredentialUnknownException("Active user is required."),
+        )
+        val clientData = if (fido2CredentialRequest.callingAppInfo.isOriginPopulated()) {
+            ClientData.DefaultWithCustomHash(
+                hash = fido2CredentialRequest.callingAppInfo.getAppSigningSignatureFingerprint(),
+            )
+        } else {
+            ClientData.DefaultWithExtraData(
+                androidPackageName = fido2CredentialRequest
+                    .callingAppInfo
+                    .getAppOrigin(),
+            )
+        }
+        val origin = fido2CredentialRequest.origin
+            ?: fido2CredentialRequest.callingAppInfo.getAppOrigin()
+
+        val ciphersWithCredentials = ciphersStateFlow.value.data
+            ?.filter { it.isActiveWithFido2Credentials }
+            .orEmpty()
+
+        val registerResult = vaultSdkSource
+            .registerFido2Credential(
+                userId = userId,
+                origin = origin,
+                requestJson = "{\"publicKey\": ${fido2CredentialRequest.requestJson}}",
+                clientData = clientData,
+                selectedCipherView = selectedCipherView,
+                cipherViews = ciphersWithCredentials,
+                isVerificationSupported = isVerificationSupported,
+                checkUser = checkUser,
+                checkUserAndPickCredentialForCreation = { options, _ ->
+                    if (options.requireVerification != Verification.DISCOURAGED) {
+                        // TODO [PM-8137]: Trigger user verification as it is preferred|required.
```

Comment:
What are your thoughts on how this will work?

Reply:
Check out #3375.